cellxgene_schema_cli/cellxgene_schema/validation_internals/check_duplicates.py

```diff
@@ -0,0 +1,81 @@
+import hashlib
+
+import anndata
+import dask.array as da
+import numpy as np
+from scipy import sparse
+from utils import get_matrix_format
+
+
+def check_duplicate_obs(adata: anndata.AnnData) -> list[str]:
+    """
+    Checks for duplicate rows in obs by raw count.
+
+    :rtype List of errors that were seen. If no errors, validation is good.
+    """
+
+    errors = []
+
+    if "in_tissue" in adata.obs.columns:
+        obs_to_keep = adata.obs[adata.obs["in_tissue"] != 0].index
+        adata = adata[obs_to_keep, :]
+
+    to_validate = [(adata.X, "adata.X")]
+
+    if adata.raw:
+        to_validate.append((adata.raw.X, "adata.raw.X"))
+
+    for matrix, matrix_name in to_validate:
+        matrix_format = get_matrix_format(matrix)
+        if matrix_format == "csr" or matrix_format == "dense":
+            errors.extend(_check_duplicate_obs_dask(adata, matrix, matrix_name))
+        else:
+            # We already raise an error for unsupported formats in _validate_sparsity, so no need to
+            # check again here.
+            continue
+
+    return errors
+
+
+def _check_duplicate_obs_dask(adata: anndata.AnnData, matrix: da.Array, matrix_name: str) -> list[str]:
+    """
+    For each row in the matrix, we compute a hash of the row and store it in row_hashes.
+    We then check for duplicates in the row_hashes.
+    """
+    errors = []
+
+    if matrix.ndim != 2:
+        return [f"Dask matrix {matrix_name} must be 2D, but got shape {matrix.shape}"]
+
+    def rowwise_hash_block(block):
+        # If it's sparse, convert each block to dense array so that we can run tobytes()
+        if sparse.issparse(block):
+            block = block.toarray()
+
+        return np.array([hashlib.sha224(row.tobytes()).hexdigest() for row in block], dtype=object).reshape(
+            -1, 1
+        )  # shape (rows, 1) for map_blocks
+
+    row_hashes = (
+        matrix.map_blocks(
+            rowwise_hash_block,
+            dtype=object,
+            chunks=(matrix.chunks[0], (1,)),  # one hash per row
+        )
+        .compute()
+        .ravel()
+    )
+
+    hash_df = adata.obs.copy()
+    hash_df["row_hash"] = row_hashes
+    dup_df = hash_df[hash_df.duplicated(subset="row_hash", keep=False)].copy()
+
+    if not dup_df.empty:
+        duplicate_rows_to_print = []
+        for i, idx in enumerate(hash_df.index[:10]):
```

Comment:
nit: can we move "10" to a var named "max_dup_rows_to_print" or something similar? to more easily track where we're setting the cutoff for how many we're printing

Reply:
also, should this be `dup_df.index`?

Reply:
yes, this should be `dup_df.index` 😬

i'll get some more feedback from lattice on what we should include in the error messages, but moving that into a separate variable seems reasonable